Cookbook for mantle convection model using tomography and other geophysical data.

alarshi opened the pull request:
This PR is a follow up from #PR 4727, which adds a cookbook to setup a mantle convection model using tomography and other geophysical data. I addressed most of the comments left on #PR 4727, but need help if the cookbook is clear enough and makes sense for a 2D spherical shell model.  Thank you!

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented:
/rebuild

gassmoeller commented on cookbooks/equilibrium_grain_size/input_data/1D_reference_profiles/prem.txt:
I think it is great to have PREM as part of ASPECT, but I am concerned we will forget that it is in this directory. Can you move the 1D_reference_profiles folder into the main data/ folder?  This way others will see it as well. This is not as important for the other data files that are more specific.

gassmoeller commented on cookbooks/equilibrium_grain_size/plugins/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+# <http://www.gnu.org/licenses/>.
+
+
+# -----------------------------------------------------------------
+# This file is a sample CMakeLists.txt file that plugins can use to
+# configure and build a shared library that can then be loaded via
+# the input file. For more information, see the manual.
+#
+# For actual use, you may want to use a different name than
+# <my_plugin> for the project, and for the names of the source files
+# below.
+# -----------------------------------------------------------------
```
Remove this part.

gassmoeller commented on doc/sphinx/user/cookbooks/geophysical-setups.md:
```diff
@@ -106,5 +106,6 @@ cookbooks/global_melt/doc/global_melt.md
 cookbooks/mid_ocean_ridge/doc/mid_ocean_ridge.md
 cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
 cookbooks/inclusions/doc/inclusions.md
+cookbooks/equilibrium_grain_size/doc/tomography_based_model.md
```
Here you name the section 'tomography based model', but the folder is called 'equilibrium grain size'. I think both are not very good names for this cookbook (tomography based model is quite unspecific, and the model has not much to do with grain size anymore). I would suggest: "Tomography based plate motions" or "tomography_based_plate_motions.prm. I know this is a lot of work, but please rename the folder and the file, and all references to the old names in all files (search and replace is your friend :-)).

alarshi replied:
Got it, thanks for suggestion!

gassmoeller reviewed (commented):
Just some first comments to let you get started, I still need to look into a lot of the files.

alarshi commented on cookbooks/equilibrium_grain_size/doc/Fig_composition.png:
Yes, that makes sense, thank you!

alarshi commented:
Thank you @gassmoeller for the review! I addressed your comments in the latest commits. We now have a different name for the model and I have added the changelog entry related to this PR.

gassmoeller commented on cookbooks/tomography_based_plate_motions/doc/tomography_based_plate_motions.md:
````diff
@@ -0,0 +1,69 @@
+```{literalinclude} input_tomography.part.txt
+```
+
+where grain_size is set to a constant value of 5mm, Vp and Vs are the P- and S-wave seismic velocities, respectively, Vs_anomaly is the S-wave velocity anomaly relative a reference velocity profile, faults represent the composition value at plate boundaries, and cratons represent the composition value within the continental cratons. Both faults and cratons have composition values set to 0 in this ...
````
```suggestion
where grain_size is set to a constant value of 5mm, Vp and Vs are the P- and S-wave seismic velocities, respectively, Vs_anomaly is the S-wave velocity anomaly relative to a reference velocity profile, faults represent the composition value at plate boundaries, and cratons represent the composition value within the continental cratons. Both faults and cratons have composition values set to 0 in this ASCII file because they are defined separately using the Geodynamics World Builder (GWB) {cite}`Fraters2019c` and the input file `cookbooks/tomography_based_plate_motions/input_file/world_builder_smac_cratons_faults_2D.json`.
```

gassmoeller commented on cookbooks/tomography_based_plate_motions/doc/tomography_based_plate_motions.part.prm:
```diff
@@ -0,0 +1,54 @@
+end
+
+subsection Material model
+  set Model name         = equilibrium grain size
```
this name should now be tomography based plate motions

gassmoeller commented on cookbooks/tomography_based_plate_motions/input_data/LLNL_model_cropped_cratons_faults.txt:
This file is very big. Please pack it into a .tar.gz archive and commit that instead (ASPECT should be able to read .tar.gz files directly). Also update the initial line of this file to correctly describe the reference it was taken from.

gassmoeller commented on cookbooks/tomography_based_plate_motions/input_data/upper_mantle_TM1_2D.txt:
```diff
@@ -0,0 +1,3244 @@
+# 2D version of the TM1 dataset, using the theta value closest to pi/2. 
```
Add a reference for TM1 here.

gassmoeller commented on cookbooks/tomography_based_plate_motions/models/gypsum_model/grain-size-test.prm:
This file is called grain-size-test, but should probably have a different name? Also the folder should not be called `models/gypsum_model/grain-size-test.prm`.

alarshi replied:
I am removing the `grain-size-test.prm` from this PR because it can be a separate test/cookbook.

gassmoeller commented on cookbooks/tomography_based_plate_motions/models/tests/2D_slice_with_faults_and_cratons.prm:
If this is a test for the cookbook it should be in the aspect/tests directory and all data paths need to be adjusted to point to tests/

alarshi replied:
Got it, thank you!

gassmoeller commented on cookbooks/tomography_based_plate_motions/plugins/reference_profile.h:
```diff
@@ -0,0 +1,207 @@
+    using namespace dealii;
+
+    /**
+     * A model in which the adiabatic profile is
+     * calculated by solving the hydrostatic equations for
+     * pressure and temperature in depth.
+     * The gravity is assumed to be in depth direction
+     * and the composition is either given by the initial
+     * composition at reference points or computed
+     * as a reference depth-function.
+     * All material parameters are computed by the
+     * material model plugin. The surface conditions are
+     * either constant or changing over time as prescribed
+     * by an user-provided function.
```
I think this is the documentation of the original ComputeProfile plugin? If so, please update with a description of what this plugin does.

gassmoeller commented on cookbooks/tomography_based_plate_motions/plugins/tomography_based_plate_motions.h:
```diff
@@ -0,0 +1,503 @@
+
+
+    /**
+     * A material model that relies on compositional fields that stand for
+     * average grain sizes of a mineral phase and source terms for them that
+     * determine the grain size evolution in dependence of the strain rate,
+     * temperature, phase transitions, and the creep regime.
+     * This material model only works if a compositional field
+     * named 'grain_size' is present. In the diffusion
+     * creep regime, the viscosity depends on this grain size. We use the grain
+     * size evolution laws described in Behn et al., 2009: "Implications of grain
+     * size evolution on the seismic structure of the oceanic upper mantle", Earth
+     * Planet. Sci. Letters, 282, 178–189. Other material parameters are either
+     * prescribed similar to the 'simple' material model, or read from data files
+     * that were generated by the Perplex or Hefesto software. The material model
+     * is described in more detail in Dannberg, J., Z. Eilon, U. Faul,
+     * R. Gassmöller, P. Moulik, and R. Myhill (2017): "The importance of
+     * grain size to mantle dynamics and seismological observations",
+     * Geochem. Geophys. Geosyst., 18, 3034–3061, doi:10.1002/2017GC006944.,
+     * which is the canonical reference for this material model.
```
I dont think this is an adequate description of this material model, it has changed a lot since description was written.

gassmoeller reviewed (commented):
Some more comments. Still havent gotten to the main files.

bangerth commented:
There are two warnings from sphinx:
```
/home/docs/checkouts/readthedocs.org/user_builds/aspect-documentation/checkouts/5256/doc/sphinx/user/cookbooks/cookbooks/tomography_based_plate_motions/doc/tomography_based_plate_motions.md:51: WARNING: Include file '/home/docs/checkouts/readthedocs.org/user_builds/aspect-documentation/checkouts/5256/doc/sphinx/user/cookbooks/cookbooks/tomography_based_plate_motions/doc/tomography_based_model.part.prm' not found or reading it failed
/home/docs/checkouts/readthedocs.org/user_builds/aspect-documentation/checkouts/5256/doc/sphinx/user/cookbooks/geophysical-setups.md:93: WARNING: toctree contains reference to nonexisting document 'user/cookbooks/<<<<<<< HEAD'
```

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/reference_profile.h:
```diff
@@ -0,0 +1,214 @@
+         */
+        Utilities::AsciiDataProfile<dim> reference_profile;
+
+        /**
+         * The column indices of the density column in the reference profile file.
+         */
+        unsigned int density_index;
```
The documentation uses the plural, but you have only one index.

bangerth commented on doc/sphinx/user/cookbooks/geophysical-setups.md:
```diff
@@ -107,7 +107,9 @@ cookbooks/global_melt/doc/global_melt.md
 cookbooks/mid_ocean_ridge/doc/mid_ocean_ridge.md
 cookbooks/kinematically_driven_subduction_2d/doc/kinematically_driven_subduction_2d.md
 cookbooks/inclusions/doc/inclusions.md
+<<<<<<< HEAD
 cookbooks/subduction_initiation/doc/subduction_initiation.md
 cookbooks/vankeken_subduction/doc/vankeken_subduction.md
+cookbooks/tomography_based_plate_motions/doc/tomography_based_plate_motions.md
```
This looks like a merge conflict. You need to remove the first line:
```suggestion
cookbooks/subduction_initiation/doc/subduction_initiation.md
cookbooks/vankeken_subduction/doc/vankeken_subduction.md
cookbooks/tomography_based_plate_motions/doc/tomography_based_plate_motions.md
```

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/reference_profile.cc:
```diff
@@ -0,0 +1,457 @@
+                                     -1;
+
+      // We want to use different values for the densities in the uppermost vs the lower part of the mantle.
+      double uppermost_mantle_thickness = 0.0;
+      if (Plugins::plugin_type_matches<const MaterialModel::TomographyBasedPlateMotions<dim>>(this->get_material_model()))
+        {
+          const MaterialModel::TomographyBasedPlateMotions<dim> &material_model
+            = Plugins::get_plugin_as_type<const MaterialModel::TomographyBasedPlateMotions<dim>> (this->get_material_model());
+
+          uppermost_mantle_thickness = material_model.get_depth_of_base_of_uppermost_mantle();
+        }
```
You initialize `uppermost_mantle_thickness` to zero if that plugin is not found. Is that the correct value? Or are you using the `0.0` value to express that the value is uninitialized and hope that you never use the value below? It seems to me the former, right?

alarshi replied:
Yes, it it the former case. If we use densities from the material model (in this case, `tomography_based_plate_motions`) , we will simply use the densities from the material model until the `uppermost_mantle_thickness` depth. We initialize it to zero so that by default the density profile is from the input reference density profile.

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/reference_profile.cc:
```diff
@@ -0,0 +1,457 @@
+            }
+          else
+            {
+              // We only want to use the PREM densities in the part of the model that also
```
Is this a copy-paste error? I thought you're not using PREM, or did I midunderstand?

alarshi replied:
Not really, we are using PREM for defining our reference profile against the tomography. But, "reference densities" may be more appropriate.

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/tomography_based_plate_motions.h:
```diff
@@ -0,0 +1,513 @@
+/*
+  Copyright (C) 2014 - 2018 by the authors of the ASPECT code.
```
```suggestion
  Copyright (C) 2023 by the authors of the ASPECT code.
```

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/tomography_based_plate_motions.cc:
```diff
@@ -0,0 +1,1726 @@
+/*
+  Copyright (C) 2014 - 2018 by the authors of the ASPECT code.
```
```suggestion
  Copyright (C) 2023 by the authors of the ASPECT code.
```

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/tomography_based_plate_motions.cc:
```diff
@@ -0,0 +1,1726 @@
+      surface_boundary_set.insert(this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"));
+      crustal_boundary_depth.initialize(surface_boundary_set, 1);
+
+      // The update() function updates the profile that stores the laterally averaged viscosity.
+      // This is needed to compute the viscosity in the material model (since the viscosities are rescaled
+      // so that the lateral average matches the reference profile). Since the viscosity depends on both
+      // temperature and velocity, we want to update the lateral average profile after each temperature
+      // and Stokes solve.
+      this->get_signals().post_stokes_solver.connect([&](const SimulatorAccess<dim> &,
+                                                         const unsigned int ,
+                                                         const unsigned int ,
+                                                         const SolverControl &,
+                                                         const SolverControl &)
+      {
+        this->update();
+      });
+
+      this->get_signals().post_advection_solver.connect([&](const SimulatorAccess<dim> &,
+                                                            const unsigned int ,
+                                                            const unsigned int ,
+                                                            const SolverControl &)
+      {
+        this->update();
+      });
```
If I understand it right, your model is instantaneous and so you never actually solve for the temperature. In that case, the comment is not quite right, nor is it necessary to connect to the `post_advection_solver`.

This is mostly for clarification to myself: You *could* update the comment and remove the second `connect()` call. But perhaps it is best to leave it in place anyway just in case someone wanted to extend our model to the time-dependent case?

bangerth commented on cookbooks/tomography_based_plate_motions/plugins/tomography_based_plate_motions.cc:
```diff
@@ -0,0 +1,1726 @@
+      for (unsigned i = 0; i < n_material_data; i++)
+        {
+          if (material_file_format == perplex)
+            material_lookup.emplace_back(std::shared_ptr<MaterialUtilities::Lookup::MaterialLookup>
+                                         (new MaterialUtilities::Lookup::PerplexReader(data_directory+material_file_names[i],
```
Please write this as follows:
```suggestion
            material_lookup.emplace_back(std::make_shared<MaterialUtilities::Lookup::PerplexReader>
                                       (data_directory+material_file_names[i],
```
same a few lines below.

bangerth reviewed (commented):
I have not looked in detail through the material model implementation, but assume that it is largely copied and pasted from other material models. 

I think I've looked through everything else, though.

alarshi commented on cookbooks/tomography_based_plate_motions/doc/tomography_based_plate_motions.md:
```diff
@@ -0,0 +1,69 @@
+
+Below `uppermost mantle thickness`, the temperature and the density fields are computed using the corresponding depth-dependent scaling values taken from{cite}`steinberger:calderwood:2006` relative to the input Vs anomalies (compositional field `Vs_anomaly`),using the scaling files `dT_vs_scaling.txt` and `rho_vs_scaling.txt`, respectively, located in `cookbooks/tomography_based_plate_motions/inp ...
+
+In order to avoid jumps in the temperature distribution, we smooth the temperatures between the two models across the `uppermost mantle thickness` depth using a sigmoid function with a half-width of 20km.
```
This is an input file parameter, but maybe I only need to reference it the first time in backticks?

bangerth replied:
That sounds reasonable.

alarshi commented on cookbooks/tomography_based_plate_motions/models/2D_slice_with_faults_and_cratons.prm:
```diff
@@ -0,0 +1,266 @@
+  subsection Stokes solver parameters
+   set Linear solver tolerance                         = 1e-4
+   set Stokes solver type                              = block GMG
+   set Number of cheap Stokes solver steps             = 5000
+   set GMRES solver restart length                     = 1000
```
No, we don't need more than 400 cheap steps. I think this is a carry over from our 3D models,  I can modify it. Thank you!

bangerth replied:
Yes, please change to 400. Let's not suggest that 1000 should be used unless one really has to.

alarshi commented on cookbooks/tomography_based_plate_motions/models/2D_slice_with_faults_and_cratons.prm:
```diff
@@ -0,0 +1,266 @@
+end
+
+
+# We use intial temperatures from an ASCII file but we also use adiabatic
+# boundary plugin such it returns very small temperatures, and on taking
```
I rephrased what I wanted to say, I hope it makes more sense now.

alarshi commented:
Thank you for the review! I have addressed all your comments, except one, in the latest commits. I will discuss the remaining one with you tomorrow.

gassmoeller commented on cookbooks/tomography_based_plate_motions/models/2D_slice_with_faults_and_cratons.prm:
```diff
@@ -0,0 +1,267 @@
+end
+
+
+# We want to use intial temperatures from an ASCII data file. However,
+# we use adiabatic boundary plugin to use the lithospheric depths but
+# set the other value of temperature gradient as zero such that on taking
+# the maximum between the two temperature models we have the temperature
+# distribution from the ASCII file.
+# We do this because we need the lithospheric depths for density computation
+# because the reference density, thermal expansivity, and compressibility varies
+# between the crust, lithosphere, and asthenosphere in our model.
```
```suggestion
# We set initial temperatures purely from an ASCII data file.
# The only reason we also activate the adiabatic boundary plugin is that it
# allows us to to read a dataset of lithospheric thicknesses that we use to
# determine reference density, thermal expansivity, and compressibility.
```

gassmoeller reviewed (commented):
Just a few typos. Afterwards ready to merge.

alarshi commented:
Thank you for the review! I addressed your comments in the latest commit and added the plugin in the `tests/` directory. Hopefully, the test works now :)

gassmoeller commented on cookbooks/tomography_based_plate_motions/models/2D_slice_with_faults_and_cratons.prm:
The tester is failing, because this .prm file is not in the base directory of the cookbook, but in the `models/` subdirectory. You can either move the .prm file there, or create a custom rule for this cookbook in cookbooks/check.mk. I would prefer moving the file into the base folder of the cookbook (also check if you have to adjust the test .prm for that.

gassmoeller reviewed (approved):
Thanks, good to go when the testers are done.

gassmoeller commented:
The relevant testers passed, so let's get this merged. Congrats @alarshi!

bangerth commented:
Yay, very nice!
